src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+    // (difference_remove emptybag B) = emptybag
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+
+  if (n[0] == n[1])
+  {
+    // (difference_remove A A) = emptybag
+    NodeManager* nm = NodeManager::currentNM();
+    Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+    return RewriteResponse(REWRITE_DONE, emptyBag);
+  }
+
+  if (n[1].getKind() == UNION_DISJOINT || n[1].getKind() == UNION_MAX)
+  {
+    if (n[0] == n[1][0] || n[0] == n[1][1])
+    {
+      // (difference_remove A (union_disjoint A B)) = emptybag
+      // (difference_remove A (union_disjoint B A)) = emptybag
+      // (difference_remove A (union_max A B)) = emptybag
+      // (difference_remove A (union_max B A)) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  if (n[0].getKind() == INTERSECTION_MIN)
+  {
+    if (n[1] == n[0][0] || n[1] == n[0][1])
+    {
+      // (difference_remove (intersection_min A B) A) = emptybag
+      // (difference_remove (intersection_min B A) A) = emptybag
+      NodeManager* nm = NodeManager::currentNM();
+      Node emptyBag = nm->mkConst(EmptyBag(n.getType()));
+      return RewriteResponse(REWRITE_DONE, emptyBag);
+    }
+  }
+
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteChoose(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_CHOOSE);
+  if (n[0].getKind() == MK_BAG && n[0][1].isConst())
+  {
+    // (bag.choose (mkBag x c)) = x where c is a constant > 0
+    return RewriteResponse(REWRITE_DONE, n[0][0]);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteCard(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_CARD);
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    // (bag.card emptybag) = 0
+    Node zero = NodeManager::currentNM()->mkConst(Rational(0));
+    return RewriteResponse(REWRITE_DONE, zero);
+  }
+  if (n[0].getKind() == MK_BAG && n[0][1].isConst())
+  {
+    // (bag.card (mkBag x c)) = c where c is a constant > 0
+    return RewriteResponse(REWRITE_DONE, n[0][1]);
+  }
+
+  if (n[0].getKind() == UNION_DISJOINT)
+  {
+    NodeManager* nm = NodeManager::currentNM();
+
+    // (bag.card (union-disjoint A B)) = (+ (bag.card A) (bag.card B))
+    Node A = nm->mkNode(BAG_CARD, n[0][0]);
+    Node B = nm->mkNode(BAG_CARD, n[0][1]);
+    Node plus = nm->mkNode(PLUS, A, B);
+    return RewriteResponse(REWRITE_AGAIN_FULL, plus);
+  }
+
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteIsSingleton(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_IS_SINGLETON);
+  NodeManager* nm = NodeManager::currentNM();
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    // (bag.is_singleton emptybag) = false
+    return RewriteResponse(REWRITE_DONE, nm->mkConst(false));
+  }
+  if (n[0].getKind() == MK_BAG)
+  {
+    // (bag.is_singleton (mkBag x c) = (c == 1)
```

Comment:
```suggestion
    // (bag.is_singleton (mkBag x c)) = (c == 1)
```

Reply:
Done.  Also removed the rule ` (bag.is_singleton emptybag)  = false` which should be in `NormalForm::getNormalForm`